### PR TITLE
8259610: VectorReshapeTests are not effective due to failing to intrinsify "VectorSupport.convert"

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch adds several c2 tests for vector reshape operations. The tests verify the intrinsification of the corresponding operations by using the IR framework and verify the correctness of the results of compiled codes.

While working on this patch, I spot some regressions regarding compilation on AVX1.

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259610](https://bugs.openjdk.java.net/browse/JDK-8259610): VectorReshapeTests are not effective due to failing to intrinsify "VectorSupport.convert"

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to 2af640c1704f6f6a5b1f176e3d8d5a7f3fda1642

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6724/head:pull/6724` \
`$ git checkout pull/6724`

Update a local copy of the PR: \
`$ git checkout pull/6724` \
`$ git pull https://git.openjdk.java.net/jdk pull/6724/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6724`

View PR using the GUI difftool: \
`$ git pr show -t 6724`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6724.diff">https://git.openjdk.java.net/jdk/pull/6724.diff</a>

</details>

### CODE REVIEW

**Comment by PaulSandoz** [comment]
Nice work. I need to look at this more closely, but this is the kind of thing i was hoping might be possible with the C2 IR test framework.

May i suggest you request to become an [author](https://openjdk.java.net/projects/#project-author)? Then you will have access to the [issue tracker](https://bugs.openjdk.java.net), and further it's one step in the process to becoming a committer, which is definitely achievable if you keep up the current rate of contributions.

**Comment by merykitty** [comment]
Hi, I'd love to. What do I need to do now?
Thank you very much.

**Comment by theRealELiu** [comment]
> Hi, I'd love to. What do I need to do now? Thank you very much.

Please refer to https://openjdk.java.net/projects/#project-author

**Review by theRealELiu — COMMENTED** [review]
Please refer to https://openjdk.java.net/projects/#project-author

File: test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastSVE.java
Comment (theRealELiu) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastSVE.java:39]: May I ask do you have physical machine to verify SVE?

Comment (merykitty) [reply test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastSVE.java:39]: I only have physical machine to verify AVX1 and AVX2. So it would be necessary to have verification from other machines, too.
Thank you very much.

**Comment by PaulSandoz** [comment]
I gave it a successful test drive (I can only run AVX1/2 locally). When looking at the code i was initially concerned over the test execution time, but it runs faster than the existing conversion tests, i think because the compilation effort is more focused.

In the run help methods, such as `runCastHelper`, you run for `INVOCATIONS = 100_000`. I am wondering if that needs to be so large if, after warmup, the method under test is compiled I believe under -Xcomp like conditions.

As a quick experiment i changed `INVOCATIONS = 1` and the casts tests passed. The IR that is generated is different (larger). IIUC the key thing is the conversion operation and species are passed as constants so the compiler can generate vector cast IR node.

**Comment by merykitty** [comment]
I have just adjusted the invocation count to be 10000 and improve test cases for floating point casts.

I have tried to reduce the invocation count further but it seems that without enough profiling c2 fails to distinguish paths, and incorrectly marks taken paths as uncommon ones. As a result, the actual operations are not compiled, which leads to IR verification failure.

To be more specific, consider the IR of a failed test as below:

    <ideal compile_id='1069'>
     177  ConL  ===  0  [[ 178 ]]  #long:8
     178  AddP  === _  10  10  177  [[ 179 ]]   Oop:java/lang/Object+8 * [narrowklass]
     190  ConNKlass  ===  0  [[ 191 ]]  #narrowklass: precise [byte: :Constant:exact *
     179  LoadNKlass  ===  87  7  178  [[ 191 ]]  @java/lang/Object+8 * [narrowklass], idx=4; #narrowklass: java/lang/Object: 0x00007ff1c8007218 *
     191  CmpN  === _  179  190  [[ 186 ]] 
     83  CmpP  === _  10  82  [[ 84 ]]  !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     151  ConL  ===  0  [[ 152 ]]  #long:12
     186  Bool  === _  191  [[ 187 ]] [ne]
     84  Bool  === _  83  [[ 85 ]] [ne] !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     5  Parm  ===  3  [[ 85 ]] Control !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     152  AddP  === _  106  106  151  [[ 153 ]]  !jvms: ByteVector::fromArray @ bci:8 (line 2902) ByteVector$ByteSpecies::fromArray @ bci:6 (line 4216) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     87  IfTrue  ===  85  [[ 187  153  179 ]] #1 !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     187  If  ===  87  186  [[ 188  189 ]] P=0.100000, C=-1.000000 !orig=[183]
     85  If  ===  5  84  [[ 86  87 ]] P=0.999999, C=-1.000000 !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     166  ConI  ===  0  [[ 168 ]]  #int:2
     153  LoadRange  ===  87  7  152  [[ 168 ]]  @bottom[int:&gt;=0]+12 * [narrow], idx=5; #int:&gt;=0 !jvms: ByteVector::fromArray @ bci:8 (line 2902) ByteVector$ByteSpecies::fromArray @ bci:6 (line 4216) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     150  ConI  ===  0  [[ 168 ]]  #int:16
     106  CheckCastPP  ===  188  10  [[ 152  152  168 ]]  #byte[int:&gt;=0]:NotNull:exact * !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     167  ConI  ===  0  [[ 168 ]]  #int:-187
     188  IfFalse  ===  187  [[ 106  168 ]] #0 !orig=[184],[175],[104] !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     10  Parm  ===  3  [[ 83  113  178  178  106 ]] Parm0: java/lang/Object *  Oop:java/lang/Object * !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     112  ConI  ===  0  [[ 113 ]]  #int:-34
     189  IfTrue  ===  187  [[ 113 ]] #1 !orig=[185],[176],[105] !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     82  ConP  ===  0  [[ 83  92 ]]  #NULL
     23  ConP  ===  0  [[ 92  168 ]]  #jdk/incubator/vector/ByteVector$ByteSpecies:exact *  Oop:jdk/incubator/vector/ByteVector$ByteSpecies:exact *
     34  ConI  ===  0  [[ 92  168 ]]  #int:0
     11  Parm  ===  3  [[ 168  92 ]] Parm1: java/lang/Object *  Oop:java/lang/Object * !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     24  ConP  ===  0  [[ 92  168 ]]  #jdk/incubator/vector/DoubleVector$DoubleSpecies:exact *  Oop:jdk/incubator/vector/DoubleVector$DoubleSpecies:exact *
     91  ConI  ===  0  [[ 92 ]]  #int:-12
     9  Parm  ===  3  [[ 113  92  168 ]] ReturnAdr !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     7  Parm  ===  3  [[ 168  153  113  92  179 ]] Memory  Memory: @BotPTR *+bot, idx=Bot; !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     6  Parm  ===  3  [[ 113  92  168 ]] I_O !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     86  IfFalse  ===  85  [[ 92 ]] #0 !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     168  CallStaticJava  ===  188  6  7  8  9 ( 167  1  1  1  24  1  11  1  1  1  1  1  1  1  1  23  106  1  1  34  150  153  166 ) [[ 169 ]] # Static uncommon_trap(reason=&apos;unstable_if&apos; action=&apos;reinterpret&apos; debug_id=&apos;0&apos;)  void ( int ) C=0.000100 VectorIntrinsics::checkFromIndexSize @ bci:3 (line 49) reexecute ByteVector::fromArray @ bci:9 (line 2902) ByteVector$ByteSpecies::fromArray @ bci:6 (line 4216) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418) !jvms: VectorIntrinsics::checkFromIndexSize @ bci:3 (line 49) ByteVector::fromArray @ bci:9 (line 2902) ByteVector$ByteSpecies::fromArray @ bci:6 (line 4216) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     113  CallStaticJava  ===  189  6  7  8  9 ( 112  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  10 ) [[ 114 ]] # Static uncommon_trap(reason=&apos;class_check&apos; action=&apos;maybe_recompile&apos; debug_id=&apos;0&apos;)  void ( int ) C=0.000100 ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418) !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     3  Start  ===  3  0  [[ 3  5  6  7  8  9  10  11 ]]  #{0:control, 1:abIO, 2:memory, 3:rawptr:BotPTR, 4:return_address, 5:java/lang/Object *, 6:java/lang/Object *}
     92  CallStaticJava  ===  86  6  7  8  9 ( 91  1  1  1  24  1  11  1  1  1  1  1  1  1  34  23  82 ) [[ 93 ]] # Static uncommon_trap(reason=&apos;null_check&apos; action=&apos;make_not_entrant&apos; debug_id=&apos;0&apos;)  void ( int ) C=0.000100 ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418) !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     169  Proj  ===  168  [[ 172 ]] #0 !jvms: VectorIntrinsics::checkFromIndexSize @ bci:3 (line 49) ByteVector::fromArray @ bci:9 (line 2902) ByteVector$ByteSpecies::fromArray @ bci:6 (line 4216) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     114  Proj  ===  113  [[ 117 ]] #0 !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     8  Parm  ===  3  [[ 168  113  117  92  96  172 ]] FramePtr !jvms: TestVectorRebracket::testB128toD128 @ bci:-1 (line 418)
     93  Proj  ===  92  [[ 96 ]] #0 !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     172  Halt  ===  169  1  1  8  1  [[ 0 ]]  !jvms: VectorIntrinsics::checkFromIndexSize @ bci:3 (line 49) ByteVector::fromArray @ bci:9 (line 2902) ByteVector$ByteSpecies::fromArray @ bci:6 (line 4216) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     117  Halt  ===  114  1  1  8  1  [[ 0 ]]  !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     96  Halt  ===  93  1  1  8  1  [[ 0 ]]  !jvms: ByteVector$ByteSpecies::fromArray @ bci:2 (line 4215) ByteVector$ByteSpecies::fromArray @ bci:3 (line 4108) VectorReshapeHelper::vectorRebracket @ bci:3 (line 256) TestVectorRebracket::testB128toD128 @ bci:8 (line 418)
     0  Root  ===  0  96  117  172  [[ 0  1  3  167  23  24  34  166  151  150  112  190  82  91  177 ]] inner 
    </ideal>

Apparently, the check cast of the parameter from an `Object` to a `byte[]` `[187]` has 2 children, the true one in which the cast is a mismatch leads to a class_check uncommon trap `[113]` while the other one leads to an unstable_if uncommon trap `[168]`. I don't know whether this is a bug, though. After compilation, the framework immediately verifies the IR before and compiled code execution, which leads to us not having a chance for recompilation.

**Comment by PaulSandoz** [comment]
I see. In this case it looks particular to rebracketing, since the test methods accept Object for the input/output. I wonder if we can simplify the rebracketing test method, at least for the output by returning the result vector or that of `Vector.toArray()`? I'll try experimenting.

**Comment by merykitty** [comment]
I have concretised the arguments of the compiled method and delegated the casts to the method handles. This gets rid of the check cast in the methods. However this time the null check has both paths leading to uncommon traps. So the invocation count still needs to be 10000 now.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastNeon.java
Comment (theRealELiu) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastNeon.java:38]: It should be `@requires vm.cpu.features ~= ".*simd.*"` 

I tested on SVE machine with MaxVectorSize=(16, 32, 64), all cases are passed. NEON passed as well after changing the pattern.

Comment (merykitty) [reply test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastNeon.java]: Thanks a lot for your testing and suggestion. Fixed.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorReinterpret.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorReinterpret.java:48]:
```suggestion
    private static final List<VectorShape> SHAPE_LIST = List.of(VectorShape.values());
```

File: test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorReinterpret.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorReinterpret.java:58]: Use `collect(joining(","))` here and in other places.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java:39]: At some point i would like to explore autogenerating such files., from another Java program and a template mechanism (unfortunately we cannot use a library like Java Poet). Not now though! 

This also relates to redesigning the existing vector unit tests, moving away from a bash script to a more flexible Java program, from which we can generate unit tests and/or IR tests.

The challenge with the IR tests is knowing associated IR nodes and the set supported on various platforms. I like how you enumerated the list for the conversions. Ideally we could go to the source of truth in C2 and determine that, but it does not seem easy to determine.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/utils/TestCastMethods.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/utils/TestCastMethods.java:42]: May later we consider a negative test that we don't expect the required IR node(s). I am uncertain in these cases whether it's an implementation restriction or a hardware restriction, if the former when the restriction goes away the test will fail and we update it.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java:296]: When `MemorySegment` previews we can remove the use of unsafe accesses, since we can view the array as a segment. Same i think also applies to direct VH accessors.

**Review by PaulSandoz — COMMENTED** [review]
Very nice work. Just a few comments.

**Comment by vnkozlov** [comment]
@chhagedorn please, review these new IR framework tests.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastAVX1.java
Comment (merykitty) [reply test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastAVX1.java]: Done, I created a helper method in `VectorReshapeHelper` and pass the necessary information to it.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java
Comment (merykitty) [reply test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java:39]: I believe if we have a way to query for compiler support from Java the same way as we do in C2 then the task would be trivial. I don't know if we have a way to call into C2 from Java similar to JNI, though.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorDoubleExpandShrink.java
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorDoubleExpandShrink.java:39]: You could use the equivalent attribute `failOn = REINTERPRET_NODE` for a zero count instead.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java:80]: The vector regexes for the IR matching could be moved to `compiler/lib/ir_framework/IRNode.java` to have them at a common place if other tests want to use them as well. If you move them you should add a `VECTOR_`/`VECTOR_CAST` prefix or something like that to better distinguish them from other nodes.

**Review by chhagedorn — APPROVED** [review]
Nice IR tests! They look good.

File: test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java
Comment (merykitty) [reply test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java]: Done, thank you very much.

**Review by chhagedorn — APPROVED** [review]
Update looks good, thanks!

File: test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java
Comment (PaulSandoz) [reply test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java:39]: Perhaps a good way to explore that is through the `WhiteBox` API?

File: test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastAVX512DQ.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/TestVectorCastAVX512DQ.java:38]:
```suggestion
 * @run driver compiler.vectorapi.reshape.TestVectorCastAVX512DQ
```

File: test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java
Comment (PaulSandoz) [inline test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java:294]:
```suggestion
    @IR(counts = {S2X_NODE, "1"})
```

**Review by PaulSandoz — COMMENTED** [review]
I ran the tests on a avx512dq machine and found a few minor issues, see prior comments. Once those are fixed I think its can be approved for integration.

**Comment by merykitty** [comment]
Thanks a lot for spotting those, they are fixed now.
/integrate

**Review by PaulSandoz — APPROVED** [review]

**Comment by TobiHartmann** [comment]
This caused a regression: https://bugs.openjdk.java.net/browse/JDK-8278623
@merykitty Please have a look.

**Comment by merykitty** [comment]
Hi @TobiHartmann , fix understood, I will submit a fix soon. Thanks.

**Pull request closed without merging**
